### PR TITLE
Fixes paid orders purchased with a 3DS card being stuck as pending payment when non-UPE is enabled

### LINKED ISSUE #3154: URGENT: Paid orders stuck at “pending payment”
**Describe the bug**
Orders are stuck at "pending payment" in Woocommerce even though customer has completed payment at checkout and Stripe is showing a successful payment with 200OK webhook response. This issue appears to have cropped up after the latest Stripe plugin 8.3 update.

**To Reproduce**
1. Update to latest version of Stripe plugin 8.3 
2. Customer completes payment successfully at checkout and payment is reflecting in Stripe with 3D secure authentication
3. Order in woocommerce remains in "pending payment" status even though Stripe webhooks show 200OK status (the order detail only shows paymentintent started and nothing else...)

**Expected behavior**
The order should be marked as processing after payment is successful, but is stuck at pending payment status. 

**Screenshots**
PLEASE NOTE THAT THIS PROBLEM HAS ALREADY BEEN REPORTED BY MULTIPLE USERS ON THE FORUM: 

1) https://wordpress.org/support/topic/paid-orders-get-a-waiting-for-payment-status/

2) https://wordpress.org/support/topic/payment-status-not-updating-from-stripe-webhook/

3) https://wordpress.org/support/topic/urgent-paid-orders-stuck-at-pending-payment/

4) https://wordpress.org/support/topic/stripe-is-not-updating-order-status/

**Environment (please complete the following information):**
 - WordPress Version - Latest version 6.5.3
 - WooCommerce Version - Latest version 8.9.1
 - Stripe Plugin Version - Latest version 8.3
 - Browser [e.g. chrome, safari] and Version - All browsers
 - Any other plugins installed - Woocommerce PayPal Payments

**Additional context**
I'm not a dev but this is a very serious problem as orders are not being reflected as processing in the Woocommerce dashboard even though they are successfully paid on Stripe's end. Please kindly fix this issue ASAP. Thank you

### PR DESCRIPTION
**Pull request by mattallan**
Fixes #3154 

## Changes proposed in this Pull Request:

In 8.3.0 we shipped a bug that prevents `payment_intent.succeeded` webhooks from being processed properly when the store has the legacy checkout experience setting enabled.

There's a couple of PRs behind the cause of this issue so I'll try and breakdown how it all led to this problem:
1. In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3037, I introduced some new metadata (`_stripe_payment_awaiting_action`) that we store on orders that require customer confirmation (i.e. payments using a 3DS card)
2. The purpose of this meta is to prevent WooCommerce's hold stock feature from cancelling Stripe orders where the customer still has the 3DS window open, or the customer is paying for the OXXO voucher.
3. This meta is used across both UPE and non-UPE payments.
4. In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3078, I incorrectly used this metadata to also prevent our webhook handlers from processing the payment_intent to avoid duplicate meta being stored on the order.
5. When processing non-UPE payments, I didn't realise we put a 5-minute lock on orders and rely on the incoming webhook to process/validate the payment which leads to [this `WC_Stripe_Helper::remove_payment_awaiting_action( $order );`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/eefdfbcdba7db375fd301915bf53e5beb54f8dd9/includes/class-wc-gateway-stripe.php#L925-L929) never running before the webhook is processed.
6. Because this meta still exists on the non-UPE order purchased with a 3DS, this then causes the webhook to exit early 😢 

There are a couple of ways we can approach fixing this problem:

1. For non-UPE we could make sure we're calling `remove_payment_awaiting_action()` when the customer has completed the checkout so that the webhook continues to used to process the payment.
2. Don't exit early from webhooks when an order still has `_stripe_payment_awaiting_action` and go back to using `_stripe_upe_waiting_for_redirect` instead.

In this PR I decided to go with option 2 as the original purpose of the `_stripe_payment_awaiting_action` was to prevent orders being cancelled by WooCommerce's hold stock feature and it wasn't intended to be used to prevent processing payment intent succeeded webhooks. This is done by reverting https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3078 and so to avoid reintroducing https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3076, I have made it so we're setting `_stripe_upe_waiting_for_redirect` inside our deferred intents implementation.

## Testing instructions

1. Enable the Legacy checkout experience and make sure webhooks are setup
2. While on `trunk` purchase a product using a 3DS card: `4000000000003220`
3. Complete the 3DS auth window
4. Once on the order thank-you page, note that the order is still pending payment
5. Open the stripe dashboard and [go to your webhooks](https://dashboard.stripe.com/test/webhooks) and confirm that the latest `payment_intent.succeeded` webhook for the order has been sent.
6. View the order in WooCommerce to confirm it's still stuck with pending payment status
7. Checkout this branch `fix/3154-paid-orders-stuck-as-pending-payment`
8. Add another product to your cart and use a 3DS card again to complete payment: `4000000000003220`
9. Once on the order thank-you page, view the order and confirm it's now set to processing/completed.

#### Other tests
 
Since this PR is reverting https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3078 and fixing it another way, we should double-check the original issue is still fixed:

1. Enable UPE and have webhooks setup
2. Purchase a product using a 3DS card again: `4000000000003220`
3. On completing the checkout, copy the order ID and search this order ID in `wp_wc_orders_meta` table (or `wp_postmeta` if you don't have HPOS enabled)
4. Confirm there's no duplicate stripe meta (i.e. `_stripe_fee` etc)

For the sake of confirming the bug, you can comment out the new line `$order->update_meta_data( '_stripe_upe_waiting_for_redirect', true );` inside `process_payment_with_deferred_intent ()` and run the test again and you will see duplicated stripe meta stored on the order.

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

**Review by james-allan — APPROVED** [review]
Changes look good to me @mattallan. 

I've verified that orders paid with a 3DS card on legacy UPE remain in a pending payment status on `develop` with the following log entry in Stripe logs: 

```
Stripe UPE waiting for redirect. The status for order 7483 might need manual adjustment.
```

I've confirmed that on this branch the order is processed via the webhook correctly.

I verified that on this branch without `$order->update_meta_data( '_stripe_upe_waiting_for_redirect', true );`, the meta is duplicated. 

Screenshot 2024-05-29 at 12 41 42 PM

With that `_stripe_upe_waiting_for_redirect` being set, that meta isn't duplicated. 

---- 

To summarise these changes in this PR: 

- It puts the new UPE in line with the legacy UPE by setting the `_stripe_upe_waiting_for_redirect` meta.
- It removes the `PAYMENT_AWAITING_ACTION_META` check in webhook processing now that both UPE modes set a consistent meta.
- `PAYMENT_AWAITING_ACTION_META` remains purely to prevent WC cancelling orders we expect are waiting for further action (eg Oxxo)

**Comment by mattallan** [comment]
Thanks @james-allan for the review and thorough testing! Merging this now to make sure it makes it into the planned patch release 🚢
